src/odemis/driver/actuator.py

```diff
@@ -2806,3 +2806,332 @@ def _doReference(self, future, axes):
+         - must stop event: to signal the movement should be stopped
+         - moving lock: to protect the sub movement during running
+        Return (CancellableFuture): a future that can be used to manage a move
+        """
+        f = CancellableFuture()
+        f._moving_lock = threading.Lock()  # taken while moving
+        f._must_stop = threading.Event()  # cancel of the current future requested
+        f._running_subf = None
+        f.task_canceller = self._cancelCurrentMove
+        return f
+
+    def _cancelCurrentMove(self, future):
+        """
+        Cancels the current move (both absolute or relative) or current reference
+        future (Future): the future to stop.
+        return (bool): True if it successfully cancelled (stopped) the move.
+        """
+        logging.debug("Cancelling current move...")
+        future._must_stop.set()  # tell the thread taking care of the move it's over
+        with future._moving_lock:
+            if future._state == FINISHED:
+                return False
+            # If future has and underlying move it should be cancelled
+            if future._running_subf:
+                future._running_subf.cancel()
+        return True
+
+    def _doMoveAbs(self, future, pos):
+        """
+        Move to the requested absolute position
+        """
+        self._adjustDepMovement(future, pos, rel=False)
+
+    def _doMoveRel(self, future, shift):
+        """
+        Move to the requested relative position
+        """
+        # Check resultant rel move still in range
+        for key in shift.keys():
+            if not self._isInRange(key, self.position.value[key] + shift[key]):
+                raise ValueError("Relative movement would go outside of range.")
+
+        self._adjustDepMovement(future, shift, rel=True)
+
+    def _isInRange(self, axis, pos=None):
+        """
+        A helper function to check if current position is in axis range
+        :param pos: if None current position is taken
+        :return: True if position in range, False otherwise
+        """
+        pos = self.position.value[axis] if pos is None else pos
+        rng = self._axes[axis].range
+        # Add 1% margin for hardware slight errors
+        margin = (rng[1] - rng[0]) * 0.01
+        return (rng[0] - margin) <= pos <= (rng[1] + margin)
+
+    def _computeLinkedAxes(self, dep_pos):
+        """
+        Compute the axes positions by mapping from the dependant stage position
+        :param dep_pos: The underlying dependant stage position
+        :return: x,y calculated position values
+        """
+        x_ = dep_pos['x']
+        y_ = dep_pos['y']
+
+        M, N, O = self._metadata[model.MD_POS_COR]
+        a, b = self._metadata[model.MD_CALIB][0]
+        c, d = self._metadata[model.MD_CALIB][1]
+
+        x = ((x_ - M) * d - (y_ - N) * b) / (a * d - b * c)
+        y = ((y_ - N) * a - (x_ - M) * c) / (a * d - b * c)
```

Comment:
Would be nice if you can make it more explicit and comment on what is lab frame (or stage frame) and what is shuttle frame. Just the underscore makes that hard to understand

Reply:
I'm also not super found of the underscore. Maybe instead of `x_`, use `xd`? 
Here, you could add a small comment indicating that it'd derived from the defining formula in `_computeDepAxes`, and assuming that Z is "at the good place".